metagraph/src/graph/annotated_graph_algorithm.cpp

```diff
@@ -1,252 +1,505 @@
+
+    // counts is a double-width, interleaved vector where the significant bits
+    // represent the out-label count and the least significant bits represent
+    // the in-label count
+    uint32_t width = counts.width() / 2;
+    uint64_t int_mask = (uint64_t(1) << width) - 1;
+
+    auto masked_graph = make_initial_masked_graph(graph_ptr,
+                                                  counts, std::move(union_mask),
+                                                  config.add_complement, MAKE_BOSS,
+                                                  num_threads);
+
+    // Filter unitigs from masked graph based on filtration criteria
+    logger->trace("Filtering out background");
+
+    tsl::hopscotch_set<std::string> masked_labels;
+    tsl::hopscotch_set<std::string> labels_in_post_set(labels_in_post.begin(),
+                                                       labels_in_post.end());
+    tsl::hopscotch_set<std::string> labels_out_post_set(labels_out_post.begin(),
+                                                        labels_out_post.end());
+
+    if (config.label_mask_other_unitig_fraction != 1.0) {
+        masked_labels.insert(labels_in.begin(), labels_in.end());
+        masked_labels.insert(labels_out.begin(), labels_out.end());
+    } else if (config.label_mask_in_unitig_fraction == 0.0
+            && config.label_mask_out_unitig_fraction == 1.0
+            && labels_in_post.empty() && labels_out_post.empty()) {
+        if (config.label_mask_in_kmer_fraction == 0.0
+                && config.label_mask_out_kmer_fraction == 1.0) {
+            logger->trace("Bypassing background filtration");
+            return std::move(*masked_graph);
+        }
 
-    const bool atomic = num_threads > 1;
+        logger->trace("Filtering by node");
+        update_masked_graph_by_node(*masked_graph, [&](node_index node) {
+            uint64_t count = count_vector.first[node];
+            uint64_t in_count = count & int_mask;
+            uint64_t out_count = count >> width;
+            uint64_t sum = in_count + out_count;
+            return in_count >= config.label_mask_in_kmer_fraction * sum
+                && out_count <= config.label_mask_out_kmer_fraction * sum;
+        }, num_threads);
+
+        return std::move(*masked_graph);
+    }
 
-    graph.call_unitigs([&](const std::string &unitig, const auto &path) {
-        if (keep_unitig(unitig, path)) {
-            for (DeBruijnGraph::node_index node : path) {
-                set_bit(unitig_mask.data(), node, atomic);
+    logger->trace("Filtering by unitig");
+
+    size_t num_labels = anno_graph.get_annotation().num_labels();
+    bool check_other = config.label_mask_other_unitig_fraction != 1.0;
+    counts.width(width);
+
+    update_masked_graph_by_unitig(*masked_graph,
+                                  [&](const auto &unitig, const auto &path)
+                                      -> std::vector<std::pair<size_t, size_t>> {
```

Comment:
This lambda is huge, so better define a function first and then call `update_masked_graph_by_unitig`.

```c++
auto ... = [&](...) {

};

update_masked_graph_by_unitig(*masked_graph, ...);